mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+  for (unsigned iterCount : this->loopIterationCounts)
+    cost *= iterCount;
+
+  // Check for temporal reuse.
+  long double temporalReuse = 1;
+
+  for (auto &accessMatrixOpPair : this->loopAccessMatrices) {
+    temporalReuse = 1;
+    SmallVector<SmallVector<int64_t, 4>, 4> accessMatrix =
+        accessMatrixOpPair.second;
+    for (int i = permutation.size() - 1; i >= 0; i--) {
+      if (!checkColumnIsZero(accessMatrix, permutation[i]))
+        break;
+      temporalReuse *= this->loopIterationCounts[permutation[i]];
+    }
+    // Increase in temporalReuse decreases the cost.
+    cost -= temporalReuse;
+  }
+  return cost;
+}
+
+/// Calculates a representative cost of this permutation regarding spatial
+/// locality. Lower cost implies better spatial reuse. Use the fact that
+/// loops with smaller locality at inner positions promote more reuse.
+long double getSpatialLocalityCost(
+    ArrayRef<unsigned> perm,
+    SmallVector<AffineForOp, 4> &loopNest,
+    DenseMap<const AffineForOp *, long double> cacheLineCount,
+    ArrayRef<unsigned> loopIterCounts) {
+
+  long double auxiliaryCost = 0;
+
+  // Product of iteration count of inner loops.
+  long double iterSubSpaceSize = 1;
+
+  // Maximum cache lines accessed by any affine.for in the loop nest. Helpful in
+  // calculating sentinel.
+  unsigned maxCacheLinesAccessed = 0;
+
+  for (int i = perm.size() - 1; i >= 0; i--) {
+    // Number of cache lines accessed by this loop at index i
+    unsigned numberCacheLinesAccessed =
+        cacheLineCount[&loopNest[perm[i]]];
+    if (numberCacheLinesAccessed > maxCacheLinesAccessed)
+      maxCacheLinesAccessed = numberCacheLinesAccessed;
+
+    auxiliaryCost += numberCacheLinesAccessed * iterSubSpaceSize;
+    iterSubSpaceSize *= loopIterCounts[perm[i]];
+  }
+
+  // The optimal permutation is one in which the affine.for ops are arranged in
+  // descending order of their individual cache line access counts from left to
+  // right.
+
+  // However such a permutation will have maximum 'auxiliaryCost' value. But we
+  // want a minimum cost for such a permutation.
+
+  // To reverse the effect, we subtract the auxiliaryCost value from a sentinel
+  // value(maximum possible spatial locality cost). We define sentinel value as
+  // follows: sentinel = Sum_for_all_loops(maxCacheLineAccessed *
+  // iterSubSpaceSize of each loop).
+
+  long double sentinel = 0;
+  iterSubSpaceSize = 1;
+  for (unsigned iterCount : loopIterCounts) {
+    sentinel += maxCacheLinesAccessed * iterSubSpaceSize;
+    iterSubSpaceSize *= iterCount;
+  }
+  return sentinel - auxiliaryCost;
+}
+
+/// Returns the optimal permutation considering both - the locality cost as well
+/// as the parallelism cost. The permutation which has the lowest overall cost is
+/// returned. Returns size-0 permutation in case no valid interchange exists.
+SmallVector<unsigned, 4> LoopInterchange::getBestPermutation(
```

Comment:
Avoid returning SmallVector's - instead use an output argument appended to arguments and pass by reference.

Reply:
Modified the function behaviour. Nowhere is a SmallVector returned. Also within the function made the changes : instead of saving bestPermutation, it now saves only the index of the best Permutation.